x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/indices/recovery/SearchableSnapshotRecoveryState.java

```diff
@@ -38,6 +39,30 @@ public synchronized RecoveryState setStage(Stage stage) {
         return super.setStage(stage);
     }
 
+    @Override
+    public synchronized RecoveryState setRemoteTranslogStage() {
+        remoteTranslogSet = true;
+        super.setStage(Stage.TRANSLOG);
+        return super.setStage(Stage.FINALIZE);
+    }
+
+    @Override
+    public synchronized void validateCurrentStage(Stage expected) {
+        if (remoteTranslogSet == false) {
+            super.validateCurrentStage(expected);
+        } else {
+            final Stage stage = getStage();
+            // For small indices it's possible that pre-warming finished shortly
+            // after transitioning to FINALIZE stage
+            if (stage != Stage.FINALIZE && stage != Stage.DONE) {
```

Comment:
I think we need to also support going back to INIT in case the entire recovery fails and we end up retrying in `RecoveryResponseHandler.handleException`? At least I am not sure we cannot get into that situation.

Also, if we need the `remoteTranslogSet` boolean, we need to also reset that in that situation.

Reply:
Yes, thanks for catching that!